pkg/scheduler/core/generic_scheduler.go

```diff
@@ -51,22 +51,55 @@ func (g *genericScheduler) Schedule(ctx context.Context, placement *policyv1alph
 		return result, fmt.Errorf("no clusters available to schedule")
 	}
 
-	feasibleClusters, err := g.findClustersThatFit(ctx, g.scheduleFramework, placement, &spec.Resource, clusterInfoSnapshot)
+	// 1. calculate reserved clusters and candidate clusters
```

Comment:
Well, I don't know why we make this code so complicated. IMO, the only thing we need to do is just deleting the scheduled result of the unjoined or unready clusters. We could just make few changes in this file.

For example, now we have a workload which is scheduled to cluster A, B and C, with 5 replicas for each.

Now cluster C is unjoined, may be now we joined a new cluster D or whatever. We could delete the scheduling result of cluster C and continue to execute the following scheduling process. The plugins could help us to find all ready clusters at present, while the `ScaleScheduling` could help us to scale the 5 replicas left by cluster C. I suppose we just need to make a small revision.

Am I right? I'm not sure.

Reply:
> IMO, the only thing we need to do is just deleting the scheduled result of the unjoined or unready clusters. We could just make few changes in this file.

If we don't reserve custers, the schedule result may completely change which breaks the workload running in original clusters

> For example, now we have a workload which is scheduled to cluster A, B and C, with 5 replicas for each.
Now cluster C is unjoined, may be now we joined a new cluster D or whatever.

**If we don't reserve custers, the schedule result may be D,E and F**, is this expected?

And this is the original behavior in `FailoverSchedule`
https://github.com/karmada-io/karmada/blob/953537d1fb9f0b199564cfbe54073026bcba754c/pkg/scheduler/core/generic_scheduler.go#L300-L301

Reply:
> If we don't reserve custers, the schedule result may be D,E and F

Does the effect come from `SpreadGroup`? I may have overlooked this field. 

I was thinking a case like this.

```
spec:
  clusters:
  - name: member1
    replicas: 1
  - name: member2
    replicas: 2
  - name: member3
    replicas: 3
```

Now member3 has been unjoined and we found new clusters member4, member5, member6 are joined.

So we delete the scheduling result of member3, like this.

```
spec:
  clusters:
  - name: member1
    replicas: 1
  - name: member2
    replicas: 2
```

The result still remains member1 and member2, and the total replicas are 3 which is less than `spec.replicas=6`. The process will be the same as `ScaleSchedule` and we don't have to worry about the old result will be removed guaranteed by `ScaleSchedule`.

But I forget to take `SpreadGroup` as a consideration, does it have an effect on candidate clusters before `assignReplicas`?

Reply:
Yes we have to take `SpreadConstraint` into consideration
https://github.com/karmada-io/karmada/blob/c83dff616330b79491154bc186f796ec7697b61c/pkg/scheduler/core/generic_scheduler.go#L250-L253

Reply:
I think the best way to solve this problem is to make a revision about selecting clusters by `SpreadConstraint`, only need to remain the previous clusters and then handling `SpreadConstraint`. So the revision will not be too large and the whole process would be clean. How do you think?

Reply:
>I think the best way to solve this problem is to make a revision about selecting clusters by `SpreadConstraint`, only need to remain the previous clusters and then handling

Yes that's exactly what I did

> IMO, the only thing we need to do is just deleting the scheduled result of the unjoined or unready clusters.

It's not just deleting them, taking `Failover` as an example, we have to consider if there is sufficient available clusters that can be scheduled, if not, we must reserve the old unready cluster

Reply:
Well, I don't think there is a need to check if there are sufficient available clusters. The check does not even contain resource validation, i.e. we still don't know if the cluster has enough resources, so is it significant for us to do this check? It may cause complicated processing and judgement.

Acutually, I prefer to delete the result of unready or unjoined clusters directly. If not matched, we do deletion. And then leave the rest as the same before. If there are no more candidates, just wait and report a rescheduling failed result until candidate clusters which have enough resources appear.

Fellows, glad to hear about your ideas, maybe I'm wrong😊. @mrlihanbo @dddddai @RainbowMango

Reply:
But it is literally the previous behavior in `FailoverSchedule`, are we gonna break it?

https://github.com/karmada-io/karmada/blob/953537d1fb9f0b199564cfbe54073026bcba754c/pkg/scheduler/core/generic_scheduler.go#L323-L330

Reply:
Hi @mrlihanbo, how do you think? 4 months ago we didn't have scale scheduling, and didn't care about the cluster resource either. Now I'm a little concerned about the main scheduling process become too complicated.